### PR TITLE
Encrypt the leader's share as well

### LINKED ISSUE #69: Encrypt to the leader as well
ACME originally had this "just trust TLS" structure and then ended up with end-to-end signatures to support designs in which you might have some kind of TLS decryption device on-path (either a forward proxy on the client or a reverse proxy on the server). ISTM that similar concerns apply here, and we should just HPKE to the leader. Note that this would also simplify the protocol because you wouldn't need to have parallel structures of leader/helper and could just have:

```
PAEncryptedInputShare input_share<..2^24-1>;
```

With the convention that leader was 0.

### PR DESCRIPTION
**Pull request by cjpatton**
Based on #70 (review that first).
Closes #69.

### CODE REVIEW

File: draft-pda-protocol.md [hunk 1]
```diff
@@ -509,17 +508,16 @@ The leader's response to malformed requests is specified in
...
-any of the following happen:
-
-* the client and helper failed to establish a secure, helper-authenticated
-  channel;
-* the GET request to the helper URL failed or didn't return a valid key config;
-  or
-* the key config specifies a KEM, KDF, or AEAD algorithm the client doesn't
+message. Similarly, when a client sends a GET request to `[leader]/key_config`,
+the leader responds with status 200 and an `HpkeConfig`. The client aborts and
+alerts the leader with "unsupported helper" if any of the following happen:
```
Comment (ekr) [inline draft-pda-protocol.md:513]: Hmm.... So how does this work if the client can't connect to the leader.

File: draft-pda-protocol.md [hunk 2]
```diff
@@ -529,81 +527,79 @@ example, that clients are prohibited from talking to helpers but not the leader.
...
+is structured as follows:
 
 ~~~
-enc, context = SetupBaseS(pk, [TODO])
+struct {
+  PATask task; // Equal to PAUploadStartReq.task
+  uint64 time; // UNIX time (in seconds).
+  PAEncryptedInputShare leader_input_share;
+  PAEncryptedInputShare helper_input_share;
+} PAUploadFinishReq;
```
Comment (ekr) [inline draft-pda-protocol.md:539]: This seems like a good opportunity to just be:

```
PAEncryptedInputShare shares<...>;
```

That is like the least we can do to lay the groundwork for >1 helper.

File: draft-pda-protocol.md [hunk 3]
```diff
@@ -529,81 +527,79 @@ example, that clients are prohibited from talking to helpers but not the leader.
...
+leader_enc, leader_context = SetupBaseS(leader_pk, [TODO])
+helper_enc, helper_context = SetupBaseS(helper_pk, [TODO])
 ~~~
 
-Field `enc` encodes the helper's encapsulated HPKE context. The remainder of the
-structure contains the share itself, the structure of which is specific to the
-PA protocol. The structure of the leader share is similarly protocol specific:
+`leader_pk` and `helper_pk` are, respectively, the leader and helper public key.
+`leader_enc` and `helper_enc` are the encapsulated HPKE context of each
+aggregator. Finally, the payloads of the encrypted input shares are computed as
```
Comment (ekr) [inline draft-pda-protocol.md:585]: This is another opportunity to avoid duplication, but just talking about aggregators.

**Review by ekr — CHANGES_REQUESTED** [review]

File: draft-pda-protocol.md (hunk 2 above)
Comment (cjpatton) [reply draft-pda-protocol.md:538]: Done.

File: draft-pda-protocol.md (hunk 3 above)
Comment (cjpatton) [reply draft-pda-protocol.md]: Done.

File: draft-pda-protocol.md [hunk 4]
```diff
@@ -559,36 +559,39 @@ struct {
...
+    case prio, helper: PrioHelperInputShare;
+    case hits, leader: HitsLeaderInputShare;
+    case hits, helper: HitsHelperInputShare;
   }
 } PAInputShare;
 ~~~
 
-Next, the client generates an HPKE {{!I-D.irtf-cfrg-hpke}} context for the
-leader and helper by running
+[TODO: Not sure if the use of `aggregator_type` above is syntactically correct.]
```
Comment (ekr) [inline draft-pda-protocol.md:576]: Yeah, you can't do this in the TLS syntax. I would just say this stuff verbally
"Of the appropriate share type for the protocol and role" down on line 593 below.

**Review by ekr — APPROVED** [review]

File: draft-pda-protocol.md (hunk 4 above)
Comment (cjpatton) [reply draft-pda-protocol.md]: Removed the PAInputShare structure altogether and clarified above that the the format of the input share depends on the protocol and aggreagtor role.

File: draft-pda-protocol.md [hunk 5]
```diff
@@ -529,81 +527,69 @@ example, that clients are prohibited from talking to helpers but not the leader.
...
 ~~~
 struct {
-  PATask task;                 // Equal to PAUploadStartReq.task
-  uint64 time;                 // UNIX time (in seconds).
-  uint8 helper_hpke_config_id; // Equal to HpkeConfig.id
-  PAHelperInputShare helper_input_share;
-  PALeaderInputShare leader_input_share;
+  PATask task; // Equal to PAUploadStartReq.task
+  uint64 time; // UNIX time (in seconds).
+  PAEncryptedInputShare encrypted_input_shares<1..2^16-1>;
```
Comment (ekr) [inline draft-pda-protocol.md:537]: nit: isn't this 7..2^16-1:

1 for config id
2 each for the lengths of enc and payload
1 each for the minimum size of length and payload.

**Review by ekr — APPROVED** [review]

File: draft-pda-protocol.md (hunk 5 above)
Comment (cjpatton) [reply draft-pda-protocol.md:540]: Yeah, well, we'll need to fix up the bounds elsewhere as well. I'll add a TODO to handle this later, once the protocol gets closer to settled.

File: draft-pda-protocol.md (hunk 1 above)
Comment (cjpatton) [reply draft-pda-protocol.md:516]: What do you mean by "feedback channel"?

Comment (chris-wood) [reply draft-pda-protocol.md:516]: I assume he means the alert from client to leader. What is the leader actually supposed to do with that information?

Comment (cjpatton) [reply draft-pda-protocol.md]: Ah yes, good call. I removed the alert.

**Comment by cjpatton** [comment]
Squashed and rebased.

**Review by tgeoghegan — APPROVED** [review]

File: draft-pda-protocol.md [hunk 6]
```diff
@@ -522,78 +523,69 @@ example, that clients are prohibited from talking to helpers but not the leader.
...
+is structured as follows:
 
 ~~~
 struct {
-  PATask task;                 // Equal to PAUploadStartReq.task
-  uint64 time;                 // UNIX time (in seconds).
-  uint8 helper_hpke_config_id; // Equal to HpkeConfig.id
-  PAHelperInputShare helper_input_share;
-  PALeaderInputShare leader_input_share;
+  PATask task; // Equal to PAUploadStartReq.task
```
Comment (chris-wood) [inline draft-pda-protocol.md:531]:
```suggestion
  PATask task;
```
(There is no more UploadStartReq, right?)

File: draft-pda-protocol.md [hunk 7]
```diff
@@ -522,78 +523,69 @@ example, that clients are prohibited from talking to helpers but not the leader.
...
+payload = context.Seal(input_share, [TODO])
 ~~~
 
-Note that only the helper's share is encrypted, and that the leader share is
-sent in plaintext. (Confidentiality of the leader's share is provided by the
-server-authenticated secure channel over which the request is made.)
+where `input_share` is the aggregator's input share.
+
+[TODO: Fully specify encryption of the shares. We need to make sure we
+authenticate things like the PA parameters and the report timestamp. We need to
```
Comment (chris-wood) [inline draft-pda-protocol.md:579]: I filed #75 to follow up on this, but: what good is authenticating the timestamp if the client can just lie about it?

**Review by chris-wood — APPROVED** [review]
LGTM with #75 spun out!

File: draft-pda-protocol.md (hunk 7 above)
Comment (cjpatton) [reply draft-pda-protocol.md:579]: Authenticating the timestamp is for the client's sake. It's meant to ensure its report is only used for a single batch.

File: draft-pda-protocol.md (hunk 6 above)
Comment (cjpatton) [reply draft-pda-protocol.md:531]: There is, as of this PR.

File: draft-pda-protocol.md (hunk 7 above)
Comment (chris-wood) [reply draft-pda-protocol.md:579]: Hmm, I don't really see how it helps the client at all. Can you elaborate?

Comment (cjpatton) [reply draft-pda-protocol.md:579]: The idea was discussed a few weeks ago, but hasn't been fully fleshed out. Basically the PA parameters specify a batch window (batch_start, batch_end). To check that a report should be added to a batch, an aggregator decrypts its input share. Decryption will fail if the reported timestamp doesn't match the timestamp generated by the client. This ensures that an honest aggregator only includes an input share in a batch if the timestamp is in the batch window.

Comment (chris-wood) [reply draft-pda-protocol.md:579]: I get that, but how does this benefit the _client_? (We should just take this to #75.)

File: draft-pda-protocol.md (hunk 6 above)
Comment (chris-wood) [reply draft-pda-protocol.md:531]: I may be missing something then. I don't see PAUploadStartReq defined in this PR's diff?

File: draft-pda-protocol.md (hunk 7 above)
Comment (cjpatton) [reply draft-pda-protocol.md:579]: Done.

**Comment by cjpatton** [comment]
Rebased.

File: draft-pda-protocol.md (hunk 6 above)
Comment (cjpatton) [reply draft-pda-protocol.md:531]: Ohhh right! The message got removed, but the "upload start" flow still exists. Good call :)

**Comment by cjpatton** [comment]
Squashed.

**Pull request merged**